rfcs/0000-arrow-user-roles.md

```diff
@@ -0,0 +1,168 @@
+- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [mbta/technology-docs#0000](https://github.com/mbta/technology-docs/pull/0000)
+- Asana task: [asana link](https://app.asana.com/)
+- Status: Proposed
+
+# Summary
+
+[summary]: #summary
+
+This is a proposed method of adding user roles and permissions levels to Arrow, perhaps to be used
+by other CTD apps.
+
+# Motivation
+
+[motivation]: #motivation
+
+Arrow is currently only accessible to admins, but we want a read-only view for "viewers", and will
+likely have other roles in the future.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The user account in Arrow is a struct, `%Accounts.User{}` with fields `:id` (an email from Active
+Directory), and `:groups`, the roles they've been assigned to, via Cognito.
+
+The current request's user is always available as `:current_user` assigned in the Plug.
+
+If you want to check if the current user's role allows them to do something, you can use the
+`Arrow.Permissions` module, like so:
+
+```ex
+def show(conn, params) do
+  can_foo? = Permissions.authorized?(conn.assigns.current_user, :do_foo)
+  # ...
+end
+```
+
+If you want to just block a controller action entirely, you can use a [Phoenix Controller plug
+pipeline](https://hexdocs.pm/phoenix/Phoenix.Controller.html#module-plug-pipeline):
+
+```ex
+defmodule ArrowWeb.BarController do
+  use ArrowWeb, :controller
+
+  plug ArrowWeb.CheckPermission, :delete_bar when action in [:delete]
+
+  # ...
+end
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## User Account and Permissions
+
+A new module will be introduced to represent user accounts:
+
+```ex
+defmodule Arrow.Accounts.User do
+  defstruct [:id, groups: []]
+end
+```
+
+and one to work with permissions. This module will centralize all authorization
+questions.
+
+```ex
+defmodule Arrow.Permissions do
+  alias Arrow.Accounts.User
+
+  @required_groups [
+    delete_bar: [:admin],
+    do_foo: [:admin, :viewer],
```

Comment:
question: What about `{verb, noun}` instead of a single atom? `{:delete, :bar}`, `{:activate, :foo}`. That also opens up a pattern where you can use pattern matching to allow permissions for a class of verbs or nouns: `{_, :bar}` for all `:bar` actions, for example, or `{:read, _}` for read-only access.